src/uct/ib/rdmacm/rdmacm_cm.c

```diff
@@ -113,6 +115,245 @@ void uct_rdmacm_cm_cqs_cleanup(uct_rdmacm_cm_t *cm)
+                  ibv_get_device_guid(verbs->device), &ret);
+    if (ret == -1) {
+        ucs_error("cm %p: cannot allocate hash entry for reserved qpn context", cm);
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    if (ret == 0) {
+        /* already exists so use it */
+        ctx = kh_value(&cm->ctxs, iter);
+    } else {
+        /* Create a reserved qpn context */
+        ctx = ucs_malloc(sizeof(struct uct_rdmacm_cm_reserved_qpn_context),
+                         "reserved_qpn_context");
+        if (ctx == NULL) {
+            kh_del(uct_rdmacm_cm_qpn_contexts, &cm->ctxs, iter);
+            ucs_error("cm %p: malloc reserved qpn context failed: %m", cm);
+            return UCS_ERR_NO_MEMORY;
+        }
+
+        pthread_mutex_init(&ctx->mutex, NULL);
+        ucs_list_head_init(&ctx->blk_list);
+        uct_rdmacm_cm_query_reserved_qpn_cap(verbs, ctx);
+        uct_rdmacm_cm_query_reserved_qpn_cap_2(verbs, ctx);
+
+        kh_value(&cm->ctxs, iter) = ctx;
+    }
+
+    *ctx_p = ctx;
+    return UCS_OK;
+}
+
+ucs_status_t
+uct_rdmacm_cm_reserved_qpn_blk_alloc(uct_rdmacm_cm_t *cm,
+                                     struct ibv_context *verbs,
+                                     struct uct_rdmacm_cm_reserved_qpn_context *ctx,
+                                     struct uct_rdmacm_cm_reserved_qpn_blk **blk_p)
+{
+    ucs_status_t status = UCS_OK;
+
+#if HAVE_DECL_MLX5DV_IS_SUPPORTED
+    char in[UCT_IB_MLX5DV_ST_SZ_BYTES(create_reserved_qpn_in)]   = {};
+    char out[UCT_IB_MLX5DV_ST_SZ_BYTES(general_obj_out_cmd_hdr)] = {};
+    struct uct_rdmacm_cm_reserved_qpn_blk *blk;
+    void *attr;
+
+    if((1 << ctx->log_reserved_qpn_granularity) > (sizeof(blk->bmp) << 3)) {
+        ucs_error("reserved qpn bit map length is not enough");
+        return UCS_ERR_INVALID_PARAM;
+    };
+
+    blk = ucs_calloc(1, sizeof(*blk), "reserved_qpn_blk");
+    if (!blk) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    UCS_BITMAP_CLEAR(&blk->bmp);
+
+    attr = UCT_IB_MLX5DV_ADDR_OF(create_reserved_qpn_in, in, hdr);
+    UCT_IB_MLX5DV_SET(general_obj_in_cmd_hdr,
+         attr, opcode, UCT_IB_MLX5_CMD_OP_CREATE_GENERAL_OBJECT);
+    UCT_IB_MLX5DV_SET(general_obj_in_cmd_hdr,
+         attr, obj_type, UCT_IB_MLX5_OBJ_TYPE_RESERVED_QPN);
+    UCT_IB_MLX5DV_SET(general_obj_in_cmd_hdr,
+         attr, log_obj_range, ctx->log_reserved_qpn_granularity);
+
+    blk->obj = mlx5dv_devx_obj_create(verbs, in, sizeof(in),
+                                      out, sizeof(out));
+    if (!blk->obj) {
+        ucs_error("mlx5dv_devx_obj_create(CREATE_GENERAL_OBJECT, type=RESERVED_QPN) failed, syndrome %x: %m",
+                  UCT_IB_MLX5DV_GET(create_mkey_out, out, syndrome));
+        status = UCS_ERR_IO_ERROR;
+        goto obj_alloc_fail;
+    }
+
+    blk->first_qpn = UCT_IB_MLX5DV_GET(general_obj_out_cmd_hdr, out, obj_id);
+    blk->next_avail_qpn_offset = 0;
+
+    *blk_p = blk;
+    return UCS_OK;
+
+obj_alloc_fail:
```

Comment:
err_free_blk

Reply:
@evgeny-leksikov Fixed. I assume you meant err_alloc_blk.